src/model/layer/cudnn_softmax.cc

```diff
@@ -26,23 +26,48 @@ CudnnSoftmax::~CudnnSoftmax() {
   if (desc_ != nullptr) CUDNN_CHECK(cudnnDestroyTensorDescriptor(desc_));
 }
 
-void CudnnSoftmax::InitCudnn(size_t size, DataType dtype) {
+void CudnnSoftmax::Setup(const LayerConf &conf) {
+  Softmax::Setup(conf);
+  SoftmaxConf sft_conf = conf.softmax_conf();
+  std::string algorithm = sft_conf.algorithm();
+  std::string mode = sft_conf.mode();
+  CHECK(algorithm == "accurate" || algorithm == "fast" || algorithm == "log")
+    << "CudnnSoftmax only supports three algorithm preferences: "
+    << "accurate, fast and log.";
+  CHECK(mode == "instance" || mode == "channel")
+    << "CudnnSoftmax only supports two modes: instance and channel.";
+  if (algorithm == "accurate")
+    algorithm_ = CUDNN_SOFTMAX_ACCURATE;
+  else if (algorithm == "fast")
+    algorithm_ = CUDNN_SOFTMAX_FAST;
+  else algorithm_ = CUDNN_SOFTMAX_LOG;
+
+  if (mode == "instance")
+    mode_ = CUDNN_SOFTMAX_MODE_INSTANCE;
+  else mode_ = CUDNN_SOFTMAX_MODE_CHANNEL;
+}
+
+void CudnnSoftmax::InitCudnn(Shape shape, DataType dtype) {
   CHECK(!has_init_cudnn_);
   CUDNN_CHECK(cudnnCreateTensorDescriptor(&desc_));
 
-  CUDNN_CHECK(cudnnSetTensor4dDescriptor(
-      desc_, CUDNN_TENSOR_NCHW, GetCudnnDataType(dtype), 1, 1, 1, size));
-
-  algorithm_ = CUDNN_SOFTMAX_ACCURATE;
-  mode_ = CUDNN_SOFTMAX_MODE_INSTANCE;
+  CHECK_LE(shape.size(), 2u)
+    << "Tensor shape should range from 1 to 2D;"
+    << "otherwise, add flatten layer to transform";
+  if (shape.size() == 1u)
+    CUDNN_CHECK(cudnnSetTensor4dDescriptor( desc_,
+      CUDNN_TENSOR_NCHW, GetCudnnDataType(dtype), 1, shape[0], 1, 1));
+  else
+    CUDNN_CHECK(cudnnSetTensor4dDescriptor( desc_, CUDNN_TENSOR_NCHW,
+      GetCudnnDataType(dtype), shape[0], shape[1], 1, 1));
   has_init_cudnn_ = true;
```

Comment:
if you set desc_ as above, then the mode should be fixed to CUDNN_SOFTMAX_MODE_INSTANCE.